modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java

```diff
@@ -138,37 +139,59 @@ public FunctionRef(Class<?> expected, PainlessMethod interfaceMethod, PainlessCo
     public FunctionRef(Class<?> expected, PainlessMethod interfaceMethod, PainlessMethod delegateMethod, int numCaptures) {
         MethodType delegateMethodType = delegateMethod.methodType;
 
-        interfaceMethodName = interfaceMethod.name;
+        interfaceMethodName = interfaceMethod.javaMethod.getName();
         factoryMethodType = MethodType.methodType(expected,
                 delegateMethodType.dropParameterTypes(numCaptures, delegateMethodType.parameterCount()));
         interfaceMethodType = interfaceMethod.methodType.dropParameterTypes(0, 1);
 
-        // the Painless$Script class can be inferred if owner is null
-        if (delegateMethod.target == null) {
-            delegateClassName = CLASS_NAME;
-            isDelegateInterface = false;
-        } else if (delegateMethod.augmentation != null) {
-            delegateClassName = delegateMethod.augmentation.getName();
-            isDelegateInterface = delegateMethod.augmentation.isInterface();
-        } else {
-            delegateClassName = delegateMethod.target.getName();
-            isDelegateInterface = delegateMethod.target.isInterface();
-        }
+        delegateClassName = delegateMethod.javaMethod.getDeclaringClass().getName();
+        isDelegateInterface = delegateMethod.javaMethod.getDeclaringClass().isInterface();
 
-        if (Modifier.isStatic(delegateMethod.modifiers)) {
+        if (Modifier.isStatic(delegateMethod.javaMethod.getModifiers())) {
             delegateInvokeType = H_INVOKESTATIC;
-        } else if (delegateMethod.target.isInterface()) {
+        } else if (delegateMethod.javaMethod.getDeclaringClass().isInterface()) {
             delegateInvokeType = H_INVOKEINTERFACE;
         } else {
             delegateInvokeType = H_INVOKEVIRTUAL;
         }
 
+        delegateMethodName = delegateMethod.javaMethod.getName();
+        this.delegateMethodType = delegateMethodType.dropParameterTypes(0, numCaptures);
+
+        this.interfaceMethod = interfaceMethod;
```

Comment:
nit: can you use `this.` consistently when setting members in ctors here? It is difficult to tell what might be locals declared earlier in the method and what are members.

Reply:
Done.